lib/phoenix_live_view/lifecycle.ex

```diff
@@ -0,0 +1,218 @@
+
+  @type t :: %__MODULE__{
+          handle_event: [Hook.t()],
+          handle_info: [Hook.t()],
+          handle_params: [Hook.t()],
+          mount: [Hook.t()]
+        }
+
+  defstruct handle_event: [], handle_info: [], handle_params: [], mount: []
+
+  @doc """
+  TODO
+
+  `fun` must match the arity of the lifecycle event callback and
+  it must return either {:cont, socket} or {:halt, socket}
+  """
+  def attach_hook(%Socket{} = socket, id, stage, fun)
+      when stage in [:handle_event, :handle_info, :handle_params] do
+    hook = Hook.new!(id, stage, fun)
+    lifecycle = lifecycle(socket)
+    existing = for h <- Map.fetch!(lifecycle, stage), h.id == id, do: h
+
+    unless existing == [] do
+      raise ArgumentError, """
+      existing hook #{inspect(hook.id)} already attached on #{inspect(hook.stage)}.
+      """
+    end
+
+    update_lifecycle(socket, stage, fn hooks ->
+      hooks ++ [hook]
+    end)
+  end
+
+  def attach_hook(%Socket{}, _id, stage, _fun) do
+    raise ArgumentError, """
+    invalid lifecycle event provided to attach_hook.
+
+    Expected one of: :handle_event | :handle_info | :handle_params
+
+    Got: #{inspect(stage)}
+    """
+  end
+
+  @doc """
+  TODO
+
+  Must match the id given to attach_hook/4
+  """
+  def detach_hook(%Socket{} = socket, id, stage)
+      when stage in [:handle_event, :handle_info, :handle_params] do
+    update_lifecycle(socket, stage, fn hooks ->
+      for hook <- hooks, hook.id != id, do: hook
+    end)
+  end
+
+  def detach_hook(%Socket{}, _id, stage) do
+    raise ArgumentError, """
+    invalid lifecycle event provided to detach_hook.
+
+    Expected one of: :handle_event | :handle_info | :handle_params
+
+    Got: #{inspect(stage)}
+    """
+  end
+
+  defp lifecycle(socket) do
+    Map.fetch!(socket.private, @lifecycle)
+  end
+
+  defp update_lifecycle(socket, stage, fun) do
+    lifecycle = lifecycle(socket)
+    new_lifecycle = Map.update!(lifecycle, stage, fun)
+    put_lifecycle(socket, new_lifecycle)
+  end
+
+  defp put_lifecycle(socket, lifecycle) do
+    put_private(socket, @lifecycle, lifecycle)
+  end
+
+  defp put_private(%Socket{private: private} = socket, key, value) when is_atom(key) do
+    %{socket | private: Map.put(private, key, value)}
+  end
+
+  # Lifecycle Event API
+
+  @doc false
+  def mount(view, hooks) when is_list(hooks) do
+    Enum.reduce(hooks, %__MODULE__{}, fn id, acc ->
+      {mod, fun} =
+        case id do
+          ^view -> raise_own_mount!(view, id)
+          {^view, :mount} -> raise_own_mount!(view, id)
+          {mod, fun} when is_atom(mod) and is_atom(fun) -> {mod, fun}
+          mod when is_atom(mod) -> {mod, :mount}
+          other -> raise_bad_mount_hook!(view, other)
+        end
+
+      hook = Hook.new!(id, :mount, Function.capture(mod, fun, 3))
+      %{acc | mount: [hook | acc.mount]}
+    end)
+  end
+
+  @doc false
+  def mount(params, session, %Socket{private: %{@lifecycle => lifecycle}} = socket) do
```

Comment:
It appears this will only catch the first mount hook, but fall into `reduce_socket` for the remaining mounts, which lack the redirect check?

Reply:
We ship the closure to `reduce_socket` and it gets invoked for each mount hook. The extra `case` here raises when the redirect is invalid, otherwise we hand the result back up to where `{:halt, socket}` will end the reduction. So I think we are okay?

I will add another test to be sure :)

Reply:
ahhh I see. We are good – thanks!